libpromises/evalfunction.c

```diff
@@ -8451,6 +8451,230 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
 
 /*********************************************************************/
 
+struct IsReadableThreadData
+{
+    pthread_t thread;
+    pthread_cond_t cond;
+    pthread_mutex_t mutex;
+    const char *path;
+    FnCallResult result;
+};
+
+static void *IsReadableThreadRoutine(void *data)
+{
+    assert(data != NULL);
+    struct IsReadableThreadData *const thread_data =
+        (struct IsReadableThreadData *) data;
+
+    // Allow this thread to be canceled asynchonously
+    int ret = pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to set thread cancel type to asynchronus: %s",
+            GetErrorStrFromCode(ret));
+        thread_data->result = FnFailure();
+        return NULL;
+    }
+
+    // Give main thread time to call pthread_cond_timedwait(3)
+    ret = pthread_mutex_lock(&(thread_data->mutex));
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to lock mutex: %s",
+            GetErrorStrFromCode(ret));
+        thread_data->result = FnFailure();
+        return NULL;
+    }
+
+    // Try to read file
+    bool success;
+    char buf[1];
+    const int fd = open(thread_data->path, O_RDONLY);
+    if (fd < 0) {
+        Log(LOG_LEVEL_DEBUG, "Failed to open file '%s': %s",
+            thread_data->path, GetErrorStr());
+        success = false;
+    }
+    else if (read(fd, buf, sizeof(buf)) < 0)
+    {
+        Log(LOG_LEVEL_DEBUG, "Failed to read from file '%s': %s",
+            thread_data->path, GetErrorStr());
+        close(fd);
+        success = false;
+    }
+    else
```

Comment:
Is it a success when `read()` returns `0` here?

Reply:
Depends on how you define if a file is readable.

> Upon reading end-of-file, zero is returned.
> - read(2)

To me, that means you've successfully read an empty file. But you can probably argue that an empty file is not readable.

Reply:
Yeah, that's why I asked that question :sweat_smile:

Reply:
I agree with Lars, an empty file is readable when you read it and get to EOF.